### PR TITLE
[TACHYON-859] respect client option to pick a specific worker

### PR DESCRIPTION
**Pull request by yuzhu**
Let me know if I misinterpreted this JIRA, because it seemed rather simple..

http://tachyon.atlassian.net/projects/TACHYON/issues/TACHYON-859

### CODE REVIEW

**Comment by calvinjia** [comment]
@yuzhu Thanks for the patch, do you mind updating the upstream calls of `getOutStream` to properly pass in the appropriate locations?

File: clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java [hunk 1]
```diff
@@ -215,7 +221,7 @@ private void checkAndAdvanceBlockInStream() throws IOException {
         try {
           // TODO(calvin): Specify the location to be local.
           mCurrentCacheStream =
-              mContext.getTachyonBlockStore().getOutStream(currentBlockId, -1, null);
+              mContext.getTachyonBlockStore().getOutStream(currentBlockId, -1, mLocation.getHost());
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java]: I don't think you want this to be the host being read from right? Since presumably the data is already there? Instead probably want to only write to local.

File: clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java [hunk 2]
```diff
@@ -273,7 +279,7 @@ private void seekBlockInStream(long newPos) throws IOException {
       if (mPos % mBlockSize == 0 && mShouldCacheCurrentBlock) {
         try {
           mCurrentCacheStream =
-              mContext.getTachyonBlockStore().getOutStream(currentBlockId, -1, null);
+              mContext.getTachyonBlockStore().getOutStream(currentBlockId, -1, mLocation.getHost());
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java]: Same comment as above.

File: clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java [hunk 3]
```diff
@@ -58,6 +61,8 @@
 
   /** If the stream is closed, this can only go from false to true */
   private boolean mClosed;
+  /** The netaddress of the location that the file should write to, specified by client option */
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileInStream.java]: The location here specifies the preferred host to read the file from.

**Comment by haoyuan** [comment]
Jenkins, test this please.

**Comment by gpang** [comment]
Jenkins, test this please

File: clients/unshaded/src/main/java/tachyon/client/ClientOptions.java [hunk 4]
```diff
@@ -148,9 +148,9 @@ public UnderStorageType getUnderStorageType() {
   }
 
   /**
-   * @return the location
+   * @return the Hostname
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/ClientOptions.java]: ``the hostname``

File: clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java [hunk 5]
```diff
@@ -45,6 +45,22 @@ public RemoteBlockOutStream(long blockId, long blockSize) throws IOException {
...
+   * @param blockSize the block size
+   * @param hostname the hostname of the preferred worker
+   * @throws IOException if I/O error occurs
+   */
+  public RemoteBlockOutStream(long blockId, long blockSize, String hostname) throws IOException {
+    super(blockId, blockSize);
+    mRemoteWriter = RemoteBlockWriter.Factory.createRemoteBlockWriter(ClientContext.getConf());
+    mWorkerClient = mContext.acquireWorkerClient(hostname);
+    mRemoteWriter.open(mWorkerClient.getDataServerAddress(), mBlockId,
+        mWorkerClient.getSessionId());
```
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java]: does this need to be on a new line? (it seems like it would fit the 100 character limit)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 6]
```diff
@@ -219,6 +225,23 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
   }
 
   /**
+   * Test writing to a file and specify the location to be localhost.
+   *
+   * @throws IOException
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: provide exception case descriptions (they can be pretty generic)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 7]
```diff
@@ -219,6 +225,23 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
...
+   * @throws InterruptedException
+   */
+  @Test
+  public void writeSpecifyLocalTest() throws IOException, InterruptedException {
+    TachyonURI filePath = new TachyonURI(PathUtils.uniqPath());
+    int len = 2;
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: ``final int LENGTH = 2;``

File: clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java (hunk 5 above)
Comment (yuzhu) [inline clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java]: actually 102 characters

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 6 above)
Comment (yuzhu) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: will do

File: clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java (hunk 5 above)
Comment (jsimsa) [inline clients/unshaded/src/main/java/tachyon/client/block/RemoteBlockOutStream.java]: so close :-)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 7 above)
Comment (yuzhu) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: fixed

File: clients/unshaded/src/main/java/tachyon/client/ClientOptions.java (hunk 4 above)
Comment (yuzhu) [inline clients/unshaded/src/main/java/tachyon/client/ClientOptions.java]: fixed.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 8]
```diff
@@ -219,6 +225,22 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
...
 
   /**
+   * Test writing to a file and specify the location to be localhost.
+   *
+   * @throws IOException if file can not be opened successfully.
+   */
+  @Test
+  public void writeSpecifyLocalTest() throws IOException {
+    TachyonURI filePath = new TachyonURI(PathUtils.uniqPath());
+    final int len = 2;
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: final int LEN (use case for constants)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 9]
```diff
@@ -228,7 +250,7 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
   @Test
   public void longWriteChangesSessionId() throws IOException, InterruptedException {
     TachyonURI filePath = new TachyonURI(PathUtils.uniqPath());
-    int len = 2;
+    final int len = 2;
```
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: ditto

**Comment by jsimsa** [comment]
Looks good to me, but Jenkins is not happy.

**Comment by yuzhu** [comment]
Jenkins, test this please.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 10]
```diff
@@ -219,6 +226,22 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
...
+   * @throws IOException if file can not be opened successfully.
+   */
+  @Test
+  public void writeSpecifyLocalTest() throws IOException, TException {
+    TachyonURI filePath = new TachyonURI(PathUtils.uniqPath());
+    final int LENGTH = 2;
+    FileOutStream os = mTfs.getOutStream(filePath, sWriteLocal);
+    os.write((byte) 0);
+    os.write((byte) 1);
+    os.close();
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: It would be good to update the functionality of `TachyonFSTestUtils` to use options in a separate PR. (Created Ticket 1040)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java [hunk 11]
```diff
@@ -219,6 +226,22 @@ private void writeTest3Util(TachyonURI filePath, ClientOptions op, int len) thro
...
 
   /**
+   * Test writing to a file and specify the location to be localhost.
+   *
+   * @throws IOException if file can not be opened successfully.
+   */
+  @Test
+  public void writeSpecifyLocalTest() throws IOException, TException {
+    TachyonURI filePath = new TachyonURI(PathUtils.uniqPath());
+    final int LENGTH = 2;
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: I don't think you need the all caps here?

File: clients/unshaded/src/main/java/tachyon/client/ClientOptions.java [hunk 12]
```diff
@@ -35,8 +34,8 @@
     private TachyonStorageType mTachyonStorageType;
     /** How this operation should interact with the under storage */
     private UnderStorageType mUnderStorageType;
-    /** Worker location to execute this operation, if not possible, the operation will fail */
-    private NetAddress mLocation;
+    /** Worker location to write data, if not possible, the operation will fail */
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/ClientOptions.java]: In the current implementation, this doc is correct. Do you mind making it also work for reading data? Though we can do that in a separate PR. (Opened Ticket 1041) It would be modifying `TachyonBlockStore#getInStream` to also take location as an argument and doing the right thing there.

**Comment by calvinjia** [comment]
@yuzhu The change looks good overall to me, I opened a few follow up tickets. For testing, did you find an easy way to do it? If not we can consider it in another PR.

File: clients/unshaded/src/main/java/tachyon/client/ClientOptions.java (hunk 12 above)
Comment (yuzhu) [inline clients/unshaded/src/main/java/tachyon/client/ClientOptions.java]: Yeah, would rather do it in a separate PR.  It is less clear to me what it means to specify a location for reads though, since the specified location might not have the data we are looking for at all?

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 11 above)
Comment (yuzhu) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: @jsimsa asked for the capitalization.  Is all caps the convention for constants?

**Comment by yuzhu** [comment]
@calvinjia Did not find an easy way to do it.. should probably consider it in another PR.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 11 above)
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: I thought it is.

Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: Right, but it needs to be `static final`? https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: clients/unshaded/src/main/java/tachyon/client/ClientOptions.java (hunk 12 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/ClientOptions.java]: Good point, maybe client options is not the best way of doing it, and it should just be the location policy.

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 11 above)
Comment (jsimsa) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: I see. In that case, let's switch it back to lowercase. @yuzhu sorry about that

**Comment by calvinjia** [comment]
@yuzhu I've opened a ticket for the unit test (1043)

File: integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java (hunk 11 above)
Comment (yuzhu) [inline integration-tests/src/test/java/tachyon/client/FileOutStreamIntegrationTest.java]: Done.

**Comment by calvinjia** [comment]
LGTM

**Comment by jsimsa** [comment]
LGTM
